### PR TITLE
fixed postgres serialisation, deserialisation of decimal

### LINKED ISSUE #241: Strange conversion to PgNumeric
Recently I came across strange behavior.

This code:
```
let value = Decimal::from_str("0.000001660").unwrap();
println!("{:?}", value);
let pg = PgNumeric::from(value);
println!("{:?}", pg);

let dec = Decimal::try_from(pg).unwrap();
println!("{:?}", dec);
```

prints:
```
0.000001660
Positive { weight: -3, scale: 9, digits: [166, 0] }
0.000000166
```

Trailing zero causes this problem. Value `0.00000166` is serialized how it should be

### PR DESCRIPTION
**Pull request by serejkaaa512**
Fixes #241

### CODE REVIEW

File: src/decimal.rs [hunk 1]
```diff
@@ -382,6 +382,43 @@ impl Decimal {
...
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use rust_decimal::Decimal;
+    ///
+    /// let number = Decimal::new(1_123, 3).with_scale(6);
+    /// assert_eq!(number, Decimal::new(1_123_000, 6));
+    /// ```
+    pub fn with_scale(self, scale: u32) -> Self {
```
[the remark this answers is not in the document]
Comment (serejkaaa512) [reply src/decimal.rs]: done

File: src/decimal.rs [hunk 2]
```diff
@@ -382,6 +382,43 @@ impl Decimal {
...
+            scale if scale > unpacked.scale => {
+                let scale_diff = scale - unpacked.scale;
+                let mul = match scale_diff {
+                    0..=9 => Decimal::from_u32(POWERS_10[scale_diff as usize]).unwrap(),
+                    10..=19 => Decimal::from_u64(BIG_POWERS_10[scale_diff as usize - 10]).unwrap(),
+                    20..=MAX_PRECISION => {
+                        Decimal::from_u64(BIG_POWERS_10[0]).unwrap()
+                            * Decimal::from_u64(BIG_POWERS_10[scale_diff as usize - 20]).unwrap()
+                    }
+                    _ => Decimal::from_u64(BIG_POWERS_10[0]).unwrap() * Decimal::from_u64(BIG_POWERS_10[8]).unwrap(),
```
[the remark this answers is not in the document]
Comment (serejkaaa512) [reply src/decimal.rs]: refactored to inner_rescale

File: src/decimal.rs [hunk 3]
```diff
@@ -382,6 +382,43 @@ impl Decimal {
...
+                };
+                let mut result = self * mul;
+                let set_scale = std::cmp::min(MAX_PRECISION, scale);
+                result.set_scale(set_scale).unwrap();
+                result
+            }
+            scale if scale < unpacked.scale => self.round_dp_with_strategy(scale, RoundingStrategy::BankersRounding),
```
[the remark this answers is not in the document]
Comment (serejkaaa512) [reply src/decimal.rs]: done

File: src/postgres.rs [hunk 4]
```diff
@@ -266,10 +231,125 @@ mod diesel {
...
+            assert_eq!(dec.to_string(), "1666500000000.00000545".to_string());
+
+            let value = Decimal::from_str("8944.000000000000").unwrap();
+            let pg = PgNumeric::from(value);
+            let dec = Decimal::try_from(pg).unwrap();
+            assert_eq!(dec.to_string(), "8944.000000000000".to_string());
+        }
+
+        #[test]
+        fn test_with_scale() {
```
[the remark this answers is not in the document]
Comment (serejkaaa512) [reply src/postgres.rs]: done

File: src/postgres.rs [hunk 5]
```diff
@@ -266,10 +231,125 @@ mod diesel {
...
 
     #[cfg(test)]
-    mod tests {
+    mod pg_tests {
         use super::*;
         use std::str::FromStr;
 
+        #[test]
+        fn test_unnecessary_zeroes() {
+            let value = Decimal::from_str("0.000001660").unwrap();
```
[the remark this answers is not in the document]
Comment (serejkaaa512) [reply src/postgres.rs:240]: done

**Comment by paupino** [comment]
I think this is looking good. There are some minor adjustments I want to make in the docs, as well as making `rescale` mutable, however I will do that as part of the merge. 
Before I finalize the merge I just want to review the `rescale_to_maximum` code to make sure it's both functionally equivalent as well as performant. 
Thank you for working through this with me!

**Comment by paupino** [comment]
For the record, performance seems fairly consistent between the two version with _this_ version being slightly more performant. Future performance benchmarking should be built into the PR process wherever possible.

**Pull request merged**